enhancements/machine-api/aws-efa-networking.md

```diff
@@ -0,0 +1,253 @@
+If a user were to specify an EFA interface type with a non supported instance (eg. m5.xlarge), then the instance will
+fail to create. In this scenario, AWS rejects the RunInstance call from Machine API which causes Machine API to mark the
+Machine as `Failed`. An error message will be added to the Machine status to signal to the user that their configuration
+was invalid.
+
+> error launching instance: EFA interfaces are not supported on m5.2xlarge
+
+## Design Details
+
+### Open Questions
+
+- What support will we provide in the future?
+  - Will we be able to provide full end to end support for the required libfabric drivers?
+- How can the feature be used by end users?
+  - How do they signal to OpenShift that a particular workload should leverage the EFA feature?
+
+### Test Plan
+
+In the first phase, we are planning to just provide users the ability to create Machines with the EFA interface type.
+We will run standard regression testing on the feature but we do not expect any special additional testing to be
+performed.
+
+In the future, if/when we introduce full support for this feature, we will need to introduce specific testing to check
+that the workloads can leverage the EFA feature and that their networking throughput is increased when leveraging the
+feature.
+
+### Graduation Criteria
+
+The addition of an API field to Machine API implies that the feature is GA from the beginning, no graduation criteria
+are required.
+
+#### Dev Preview -> Tech Preview
+
+N/A
+
+#### Tech Preview -> GA
+
+N/A
+
+#### Removing a deprecated feature
+
+No features will be removed as a part of this proposal
+
+### Upgrade / Downgrade Strategy
+
+Existing clusters being upgraded will not be configured for the new feature and will continue to use the existing
+standard interface type.
+
+Once configured, on downgrade, the Machine API components will not know about the new fields, and as such, will ignore
+the interface type field if specified. The usage of an EFA interface type will not affect removal of Machines after a
+downgrade, there should be no detrimental effect of a downgrade on Machine API.
+
+### Version Skew Strategy
+
+N/A
+
+### Operational Aspects of API Extensions
+
+N/A
+
+#### Failure Modes
+
+N/A
+
+#### Support Procedures
+
+N/A
+
+## Implementation History
+
+The Machine API changes are implemented within:
+- https://github.com/openshift/api/pull/1065
+- https://github.com/openshift/machine-api-provider-aws/pull/8
+
+## Drawbacks
+
+This feature is not easily usable by a generic user. There are a number of risks outlined in the risk section that
+provide reasons for not implementing this feature, to summarise:
+- End users must modify security group rules before they can use this feature
+- End users must deploy the libfabric software onto OpenShift before they can use this feature
+- We cannot supply this feature for control plane machines without replacing the control plane machines as a day 2
+  operation
+
+## Alternatives
```

Comment:
Another alternative would be to provide more explicit high-level support for HPC. Eg, create a "cluster-hpc-operator" that can configure a MachinePool for HPC computing in a manner appropriate for the platform.

On AWS this would mean requesting EFAs instead of ENAs, which might also require using a different instance type than the worker pool uses. On Azure this would mean using [RDMA-capable instances](https://docs.microsoft.com/en-us/azure/virtual-machines/sizes-hpc#rdma-capable-instances), which would have an HPC-enabled interface as a _secondary_ network interface.

On all platforms this would involve setting placement groups/policy of the VMs, to ensure that all the nodes in the HPC MachinePool are running on physically-close hardware. GCP also has [lots of other recommendations](https://cloud.google.com/architecture/best-practices-for-using-mpi-on-compute-engine) about how to configure HPC VMs.

Assuming we eventually come up with the APIs for it, the operator could also manage the cloud security groups. (The can't-have-a-firewall thing is not AWS-specific, it's a requirement of MPI, which uses arbitrary random ports for communication.)

Reply:
Im not sure why is it an alternative. This will be future enhancement, but in current time frame, I dont see how we can pursue this path.
Hence we should limit this enhancement only to screen what are the current expectations and deliverable are and later on we'll improve / adjust.